File: ios/Runner/MediaPipe/JpegEncoderBridge.swift
import Foundation
import UIKit

/// Encodes a `FrameArgs` payload as an upright JPEG written to `outPath`.
///
/// Reuses `FrameDecoder.decodeUprightImage` (BGRA / RGBA / YUV → upright UIImage)
/// and emits a JPEG at the requested quality. Returns the path on success or
/// `nil` if decoding/encoding failed.
enum JpegEncoderBridge {
    static func encode(frame: FrameArgs, quality: Int, outPath: String) -> String? {
        guard let image = FrameDecoder.decodeUprightImage(frame: frame) else {
            return nil
        }
        let q = max(1, min(100, quality))
        let cq = CGFloat(q) / 100.0
        guard let data = image.jpegData(compressionQuality: cq) else { return nil }
        do {
            try data.write(to: URL(fileURLWithPath: outPath), options: .atomic)
            return outPath
        } catch {
            return nil
        }
    }
}

File: android/app/src/main/kotlin/com/example/face_liveness_demo/mediapipe/JpegEncoderBridge.kt
package com.example.face_liveness_demo.mediapipe

import android.graphics.Bitmap
import java.io.File
import java.io.FileOutputStream

/**
 * Encodes a [FrameArgs] payload as an upright JPEG written to [outPath].
 *
 * Reuses [FrameDecoder.decodeUprightBitmap] (NV21 / YUV / BGRA / RGBA → upright RGB
 * Bitmap) and pipes the result through [Bitmap.compress] at the given quality.
 *
 * Returns the absolute path on success, or null if decode/compression failed.
 */
object JpegEncoderBridge {
    fun encode(frame: FrameArgs, quality: Int, outPath: String): String? {
        val bitmap = FrameDecoder.decodeUprightBitmap(frame) ?: return null
        try {
            FileOutputStream(File(outPath)).use { fos ->
                if (!bitmap.compress(Bitmap.CompressFormat.JPEG, quality, fos)) {
                    return null
                }
                fos.flush()
            }
        } finally {
            bitmap.recycle()
        }
        return outPath
    }
}
